### ISSUE TITLE
Investigate text to speech models

### ISSUE THREAD

**Issue by nchapman** [nchapman/nebula#12]
We need to choose a text to speech model to integrate. The [TTS Arena](https://huggingface.co/spaces/TTS-AGI/TTS-Arena) has a good list for us to consider:

![CleanShot 2024-03-24 at 20 49 12@2x](https://github.com/nchapman/nebula/assets/3095/81497337-fd9a-4b6b-8935-32aa8e9175ac)

[StyleTTS2](https://github.com/yl4579/StyleTTS2) and [XTTS2](https://huggingface.co/coqui/XTTS-v2) seem most promising.

We need to figure out which of these performs best (especially for their size) and which would be eaiest to integrate into Nebula.

**Comment by iKachaykin** [comment]
@nchapman Great, thanks for this reference! Will consider it as well 👍

**Comment by iKachaykin** [comment]
@nchapman We have the following status here so far:

- All the models mentioned in the table are implemented with pure Python and PyTorch.
- In such a way there is no easy and straightforward way to run them directly from Rust, for sure if it isn't an option to just invoke some Python API that we host on the cloud, or even in a local HTTP server. (We didn't discuss it, but I believe that's not a solution for us, right? (1) )
- Having said that, there are some ways how we may integrate text-to-speech into nebula (approaches figured out at the moment, maybe some other exist, please let me know if I should do more research regarding possible options (2) ):
    1. Integrate other models that are implemented directly in Rust. Some examples: [MetaVoice](https://github.com/huggingface/candle/tree/main/candle-examples/examples/metavoice) available in `candle` crate, or [sonata](https://github.com/mush42/sonata/tree/main) that is Rust bindings of [Piper](https://github.com/rhasspy/piper) model. These models are not state-of-the-art for sure, but it may be a good start to have some TTS models available in nebula.
    2. For the models implemented with Python and PyTorch, decompose them into bricks, convert as much bricks as possible into [Torch Script (.pt)](https://pytorch.org/docs/stable/jit.html) files, then run .pt files in Rust using the following package: [tch-rs](https://github.com/LaurentMazare/tch-rs) as described [here](https://github.com/LaurentMazare/tch-rs/blob/main/examples/jit/README.md), after that finally develop all the missing logic (mainly infrastructure, i.e. weights automatic download, data loading, writing into files, etc.) with Rust and suitable crates. This approach will for sure require some re-implementation to do, **but** it doesn't require to re-implement everything that is a strong advantage of this approach and may be a good alternative to re-implement everything from scratch. This approach has some limitations as well however, as only low-level PyTorch `nn.Module` objects may be converted into .pt it seems. I had a hope that we have some way to convert high-level models as well, but for now I am a bit skeptical about that as I still didn't succeed to achieve that. Anyway, it should be easier than re-implementing everything from scratch. P.S.: I include into this approach sub-approaches like convert models to `onnx`, and run them from Rust using [onnxruntime](https://docs.rs/onnxruntime/latest/onnxruntime/).
    3. The next and brute-force option consists in re-implementing all/some of these models from scratch with `candle`. On one hand, that is always the most direct and straightforward option that is always possible to do, and that will always work. Fortunately, a lot of ready-to-use elements are available within `candle`, so we may reuse them on development. On another hand, it will require a lot of development effort (= time). Another issue that I see at this moment: we will have to maintain much more code than with other options. Finally, if these models are no more the state-of-the-art, then we will have to reiterate once again and implement new models, that will take again more time than with other options if ever (or rather when) this situation takes place. A little note for this approach: it will maybe make sense to re-implement everything in C++ if we end up with this solution, and then create bindings for Rust instead of implementing directly with Rust. Reason: C++ implementation will be easier to reuse in other packages and projects if needed.
    4. Final approach that I personally think will work worse than others, but anyway it is worth to discuss. Using the [pyo3 repo](https://github.com/PyO3/pyo3), create bindings for Rust directly from Python. I am not sure if it is a good solution, as it will require to have Python installed and will literally run Python on the background. So we will have all the classic performance and other issues. And it looks like over-engineering to be honest: if we end-up running Python on the background, why don't we just use Python API instead? So this solution is not a good choice for us I believe. Anyway it is worth to mention it, as you may find it interesting.

Please let me know what do you think about these results overall, and in particular please let me know what is your feedback about aspects (1) and (2) :)

Otherwise, I am continuing my research about approach **ii** to figure out the highest level of the abstraction acceptable to be converted to a .pt file. On the low level everything will work, but what will be the optimal level of abstraction for us, so that we may do as less job as possible while integrating into Rust? My goal is to answer on this question. But please let me know if I should concentrate on something else, as you definitely see the overall priorities much better :)

**Comment by nchapman** [comment]
This is a great write up. Thank you @iKachaykin! I agree. Let's prototype option `ii` during this sprint and see if we can get a working prototype. I am also very curious about onnx. What do you think about prototyping that option as well? I think it's worth spending some extra time on this solution since we'd be able to reuse it for other pytorch models we might want to integrate in the future.
